### ISSUE TITLE
VTK compilation error (buggy openGL lib problem) on Ubuntu AND Fedora Core

### ISSUE THREAD

**Issue by jypeter** [UV-CDAT/uvcdat#983]
I have had on Fedora Core (FC20)  the same VTK/openGL related compilation problem that was already reported for Ubuntu

The installation fails with the following error
```
Performing build step for 'VTK'
CMake Error at cdat_build_uv-2.1.0_x86_64_gcc4_VB_13/VTK-prefix/src/VTK-stamp/VTK-build.cmake:56 (message):
  CTEST_FULL_OUTPUT

   Command failed: 2
   'make'
   From cdat_build_uv-2.1.0_x86_64_gcc4_VB_13/VTK-prefix/src/VTK-stamp/VTK-build-out.log
[...]
   From cdat_build_uv-2.1.0_x86_64_gcc4_VB_13/VTK-prefix/src/VTK-stamp/VTK-build-err.log
[...]
```
and VTK-build-err.log reports several errors of the following type
```
In file included from /usr/include/GL/glx.h:333:0,
                 from cdat_build_uv-2.1.0_x86_64_gcc4_VB_13/build/VTK/Rendering/OpenGL/vtkXOpenGLRenderWindow.cxx:30:
/usr/include/GL/glxext.h:480:143: error: 'GLintptr' has not been declared
 typedef void ( *PFNGLXCOPYBUFFERSUBDATANVPROC) (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLenum readTarget, GLenum writeTarget, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);

/usr/include/GL/glxext.h:480:164: error: 'GLintptr' has not been declared
 typedef void ( *PFNGLXCOPYBUFFERSUBDATANVPROC) (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLenum readTarget, GLenum writeTarget, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
                                                                                                                                                                    ^
/usr/include/GL/glxext.h:480:186: error: 'GLsizeiptr' has not been declared
 typedef void ( *PFNGLXCOPYBUFFERSUBDATANVPROC) (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLenum readTarget, GLenum writeTarget, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
```
As was already reported for Ubuntu, it seems that the compilation problems occur on **too recent** versions of the mesa/GL library. The compilation went fine on our RH EL server

    A) Red Hat Enterprise Linux Server release 6.6 (Santiago)

but failed on my Virtual Machine

    B) Fedora release 20 (Heisenbug)

The libraries are indeed slightly more recent on the VM
```
rpm -qf /usr/include/GL/glxext.h
	A) mesa-libGL-devel-10.1.2-2.el6.x86_64
	B) mesa-libGL-devel-10.3.3-1.20141110.fc20.x86_64
```

The solution given by @aashish24 on the wiki does work but I find it awful to have to edit an installed library in the system as root! By chance, I could be root on my own VM, but I would not been able to do that on a server. The solution is to add the typedefs for *GLsizeiptr* and *GLintptr* in **/usr/include/GL/glxext.h** and then type **make** again in the build directory
```
#endif /* GLX_MESA_set_3dfx_mode */

#ifndef GLX_NV_copy_buffer
#define GLX_NV_copy_buffer 1

/* Jan 27th 2015 : Added the 2 typedef below
   in order to be able to compile VTK and install UV-CDAT */
typedef ptrdiff_t GLsizeiptr;
typedef ptrdiff_t GLintptr;

typedef void ( *PFNGLXCOPYBUFFERSUBDATANVPROC) (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLenum readTarget, GLenum writeTarget, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
typedef void ( *PFNGLXNAMEDCOPYBUFFERSUBDATANVPROC) (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLuint readBuffer, GLuint writeBuffer, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
#ifdef GLX_GLXEXT_PROTOTYPES
void glXCopyBufferSubDataNV (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLenum readTarget, GLenum writeTarget, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
void glXNamedCopyBufferSubDataNV (Display *dpy, GLXContext readCtx, GLXContext writeCtx, GLuint readBuffer, GLuint writeBuffer, GLintptr readOffset, GLintptr writeOffset, GLsizeiptr size);
#endif
#endif /* GLX_NV_copy_buffer */
```

Since the problem mostly comes from VTK using the bad include file, and UV-CDAT downloads VTK, would it be possible to patch the downloaded VTK (where we have write access) rather than do something as root? Quickly looking up this on Google showed the following https://bugzilla.redhat.com/show_bug.cgi?id=1138466 but I have not tested this

**Comment by dlonie** [comment]
On Tue, Jan 27, 2015 at 11:19 AM, Jean-Yves Peterschmitt <
notifications@github.com> wrote:

> The solution given by @aashish24 <https://github.com/aashish24> on the
> wiki does work but I find it awful to have to edit an installed library in
> the system as root! By chance, I could be root on my own VM, but I would
> not been able to do that on a server. The solution is to add the typedefs
> for *GLsizeiptr* and *GLintptr* in */usr/include/GL/glxext.h* and then
> type *make* again in the build directory
>
> Since the problem mostly comes from VTK using the bad include file, and UV-CDAT downloads VTK, would it be possible to patch the downloaded VTK (where we have write access) rather than do something as root? Quickly looking up this on Google showed the following https://bugzilla.redhat.com/show_bug.cgi?id=1138466 but I have not tested this
>
>
I agree that it's annoying to have to edit the system files, but IMO it's
the best solution at the moment. The bug is in their gl headers, and I'd
rather not start slicing out OpenGL features from VTK as that may have
unforeseen consequences.

My vote is to keep things the way they are -- the bug is at the distro
level, and they should fix it eventually (TBH, I'm not sure how it has gone
this long without getting fixed!). This won't affect non-technical users
(they can just download the binaries and not worry about modifying these
headers), and only developers under specific circumstances will need to
concern themselves with this.

Commenting out the define that brings in the bad header might work just
fine, but I'd rather leave that as an experiment for cases where the system
files cannot be modified, rather than shutting it off for everyone. If
someone is running into this issue on a system where they can't modify the
glext.h header, they would be able to try the alternative solution locally.

Just my 2c.

Dave

**Comment by dlonie** [comment]
@jbeezley pointed this out:

http://public.kitware.com/pipermail/vtkusers/2014-September/085057.html

This has actually been worked-around in VTK master already -- the issue should go away once VTK is bumped.

**Comment by doutriaux1** [comment]
nice!

**Comment by jypeter** [comment]
So, meanwhile, all we have to do is uncomment the line defining **GLX_GLXEXT_LEGACY** in 
**build/VTK/Rendering/OpenGL/vtkXOpenGLRenderWindow.cxx** before running *make*, right?

I thought it might be that, but I wonder why the line was commented out
```
// define GLX_GLXEXT_LEGACY to prevent glx.h to include glxext.h provided by
// the system
//#define GLX_GLXEXT_LEGACY
#include "GL/glx.h"
```

**Comment by dlonie** [comment]
Sounds like that should do the trick, but be sure to run the tests
afterwards ;)

**Comment by jypeter** [comment]
Which tests? Just the full uv-cdat tests by running 'ctest' in uv-cdat build dir? Anyway, hopefully I won't run into this problem on other platforms and I won't have the opportunity to try this

**Comment by dlonie** [comment]
Yes, the uvcdat tests should be enough (VTK has it's own test suite, but
I'm pretty sure it's disabled by default). To be safe, one could run both,
but just making sure that uvcdat's functionality isn't affected should be
fine.

**Comment by jypeter** [comment]
@dlonie I have re-installed 2.1.0 with a **patched vtkXOpenGLRenderWindow.cxx instead of a patched /usr/include/GL/glxext.h** and indeed all the ctest passed!
```
100% tests passed, 0 tests failed out of 343

Total Test time (real) = 862.86 sec
```

I hoped that maybe this would solve by magic my VirtualBox+FedoraCore+GL problem, but it did not. This is not directly a UV-CDAT problem, but the support of GL hardware rendering in VB is quite lousy and when I plot something in a vcs canvas, my full Virtual Machine gets blank (black actually).

I will probably open a detailed issue about that to see if somebody has an idea. I don't see the blinking vcs test windows during ctest, because my full VB blinks in sync with the tests, but the images must be composed correctly offscreen, since all the tests pass. It's just a major rendering problem in my VM :( And this means that, though uvcdat is installed correctly, vcs is unusable in this particular VM
